### PR TITLE
iam-user - add ability to pick what parts of the iam-user to delete

### PR DESCRIPTION
**Pull request by thisisshi**
related to: https://github.com/capitalone/cloud-custodian/issues/2450

New Functionality:
- Specify parts of an iam-user to delete using `options`
  - console-access
  - access-keys
  - user-policies
  - mfa-devices
  - groups
  - ssh-keys
  - signing-certificates
  - service-specific-credentials
```yaml
policies:
  - name: delete-user-console-access
    resource: iam-user
    filters:
      - UserName: test_user
    actions:
      - type: delete
        options:
          - console-access
```

- To delete the user itself, use `delete` with no options (previous implementation):

```yaml
policies:
  - name: delete-user
    resource: iam-user
    filters:
      - UserName: test_user
    actions:
      - type: delete
```

### CODE REVIEW

File: c7n/resources/iam.py [hunk 1]
```diff
@@ -1108,83 +1108,118 @@ class UserDelete(BaseAction):
...
         response = client.list_service_specific_credentials(UserName=r['UserName'])
         for screds in response.get('ServiceSpecificCredentials', ()):
             client.delete_service_specific_credential(
                 UserName=r['UserName'],
                 ServiceSpecificCredentialId=screds['ServiceSpecificCredentialId'])
 
+    def delete_user(self, client, r):
         client.delete_user(UserName=r['UserName'])
 
+    OPTIONS = {
```
Comment (kapilt) [inline c7n/resources/iam.py]: can you move the class vars above the methods in the class def

File: c7n/resources/iam.py [hunk 2]
```diff
@@ -1108,83 +1108,118 @@ class UserDelete(BaseAction):
...
+        'console-access': delete_console_access,
+        'access-keys': delete_access_keys,
+        'user-policies': delete_user_policies,
+        'mfa-devices': delete_hw_mfa_devices,
+        'groups': delete_groups,
+        'ssh-keys': delete_ssh_keys,
+        'signing-certificates': delete_signing_certificates,
+        'services-specific-credentials': delete_service_specific_credentials,
+        'user': delete_user,
```
Comment (kapilt) [inline c7n/resources/iam.py]: we should omit delete user as a config option doing it in isolation is error prone without its deps and its the default action.

File: c7n/resources/iam.py [hunk 3]
```diff
@@ -1108,83 +1108,118 @@ class UserDelete(BaseAction):
...
+        'iam:DeleteSigningCertificate',
+        'iam:DeleteSSHPublicKey',
+        'iam:DeleteUser',
+        'iam:DetachUserPolicy',
+        'iam:RemoveUserFromGroup')
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('iam')
+        for r in resources:
+            self.log.debug('Deleting user %s options: %s' %
```
Comment (kapilt) [inline c7n/resources/iam.py]: per resource logging on the happy path, is a bit verbose even at the debug level

File: c7n/resources/iam.py [hunk 4]
```diff
@@ -1108,83 +1108,118 @@ class UserDelete(BaseAction):
...
+                (r['UserName'], self.data.get('options', 'all')))
+            self.process_user(client, r)
+
+    def process_user(self, client, r):
+        options = self.data.get('options', list(self.OPTIONS.keys()))
```
Comment (kapilt) [inline c7n/resources/iam.py]: in python2 dictionary ordering is random, so you could get the remove user first which would error out due to dependencies, the options dict needs to be an ordered dict to handle the order. it looks like your already handling this by hardcoding the 'user' option and then removing it separately, but using an ordered dict would resolve this more simply.

File: c7n/resources/iam.py (hunk 1 above)
Comment (thisisshi) [reply c7n/resources/iam.py]: I originally had them above but the reference to `OPTIONS.keys` in schema required it to be below the method definitions. I'm thinking of moving the `delete_*` outside of the class itself, as `delete_access_keys` could actually be reused in the `remove-keys`

File: c7n/resources/iam.py (hunk 2 above)
Comment (thisisshi) [reply c7n/resources/iam.py]: Yeah, I was in between including/removing `user` as an option as the order of ops requires it to be last, I'll remove and change to an ordered dict and include special handling for when `options is None`

Comment (thisisshi) [reply c7n/resources/iam.py]: If we don't include `user` in `options` then supplying `options` means that we should not delete the user?

File: c7n/resources/iam.py (hunk 1 above)
Comment (thisisshi) [reply c7n/resources/iam.py]: ended up switching to use getattr and using a dict of `{str:str}` to be able to move the class vars to the top while keeping the methods inside of the class for encapsulation

**Pull request merged**